kv/kvserver: TestLeaseRenewer failed

kv/kvserver.TestLeaseRenewer [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9444080?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9444080?buildTab=artifacts#/) on master @ [ff2c4a3454a63bd74fe9a2feed53a7bdcfaf5191](https://github.com/cockroachdb/cockroach/commits/ff2c4a3454a63bd74fe9a2feed53a7bdcfaf5191):


```
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:143: r1 renewers: [1]
    replica_lease_renewal_test.go:125: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_lease_renewal_test.go:125
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_lease_renewal_test.go:165
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23
        	Error:      	Condition never satisfied
        	Test:       	TestLeaseRenewer/kv.expiration_leases_only.enabled=true
    --- FAIL: TestLeaseRenewer/kv.expiration_leases_only.enabled=true (1573.75s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseRenewer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26558

--- a/pkg/kv/kvserver/replica_lease_renewal_test.go
+++ b/pkg/kv/kvserver/replica_lease_renewal_test.go
@@ -36,7 +36,10 @@ func TestLeaseRenewer(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 100689)
+	// stressrace and deadlock make the test too slow, resulting in an inability
+	// to maintain leases and Raft leadership.
+	skip.UnderStressRace(t)
+	skip.UnderDeadlock(t)
 
 	// When kv.expiration_leases_only.enabled is true, the Raft scheduler is
 	// responsible for extensions, but we still track expiration leases for system
@@ -48,9 +51,18 @@ func TestLeaseRenewer(t *testing.T) {
 		tc := serverutils.StartNewTestCluster(t, 3, base.TestClusterArgs{
 			ServerArgs: base.TestServerArgs{
 				Settings: st,
+				// Speed up lease extensions to speed up the test, but adjust tick-based
+				// timeouts to retain their default wall-time values.
 				RaftConfig: base.RaftConfig{
-					RangeLeaseDuration: time.Second,
-					RaftTickInterval:   100 * time.Millisecond,
+					RangeLeaseRenewalFraction:  0.95,
+					RaftTickInterval:           100 * time.Millisecond,
+					RaftElectionTimeoutTicks:   20,
+					RaftReproposalTimeoutTicks: 30,
+				},
+				Knobs: base.TestingKnobs{
+					Store: &StoreTestingKnobs{
+						LeaseRenewalDurationOverride: 100 * time.Millisecond,
+					},
 				},
 			},
 		})
@@ -173,7 +185,8 @@ func TestLeaseRenewer(t *testing.T) {
 		assertStoreLeaseRenewer(desc.RangeID)
 
 		// Transfer the lease to a different leaseholder, and assert that the lease is
-		// still extended.
+		// still extended. Wait for the split to apply on all nodes first.
+		require.NoError(t, tc.WaitForFullReplication())
 		lease, _ := getNodeReplica(1, desc.RangeID).GetLease()
 		target := tc.Target(lookupNode(lease.Replica.NodeID%3 + 1))
 		tc.TransferRangeLeaseOrFatal(t, desc, target)

--- a/pkg/kv/kvserver/replica_proposal.go
+++ b/pkg/kv/kvserver/replica_proposal.go
@@ -372,8 +372,8 @@ func (r *Replica) leasePostApplyLocked(
 		r.gossipFirstRangeLocked(ctx)
 	}
 
-	if newLease.Type() == roachpb.LeaseExpiration && (leaseChangingHands || maybeSplit) &&
-		iAmTheLeaseHolder && r.ownsValidLeaseRLocked(ctx, now) {
+	isExpirationLease := newLease.Type() == roachpb.LeaseExpiration
+	if isExpirationLease && (leaseChangingHands || maybeSplit) && iAmTheLeaseHolder {
 		if r.requiresExpirationLeaseRLocked() {
 			// Whenever we first acquire an expiration-based lease for a range that
 			// requires it (i.e. the liveness or meta ranges), notify the lease
@@ -388,7 +388,7 @@ func (r *Replica) leasePostApplyLocked(
 			case r.store.renewableLeasesSignal <- struct{}{}:
 			default:
 			}
-		} else if !r.shouldUseExpirationLeaseRLocked() {
+		} else if !r.shouldUseExpirationLeaseRLocked() && r.ownsValidLeaseRLocked(ctx, now) {
 			// We received an expiration lease for a range that shouldn't keep using
 			// it, most likely as part of a lease transfer (which is always
 			// expiration-based). We've also applied it before it has expired. Upgrade

